backend/src/org/commcare/xml/ProfileParser.java

```diff
@@ -48,46 +47,57 @@ public Profile parse() throws InvalidStructureException, IOException, XmlPullPar
-        if(sMajor != null) {
+        if (sMajor != null) {
             major = parseInt(sMajor);
         }
         
-        if(sMinor != null) {
+        if (sMinor != null) {
             minor = parseInt(sMinor);
         }
 
-        //If version information is available, check valid versions
-        if ((!forceVersion && this.instance != null) && (major != -1) && (minor != -1)){
+        // If version information is available, check valid versions
+        if ((!forceVersion && this.instance != null) && (major != -1) && (minor != -1)) {
             
-            //For the major version, only a matching number is valid, 2.0 cannot be run on either 1.0 or 3.0
+            // For the major version, only a matching number is valid, 2.0 cannot be run on either 1.0 or 3.0
             if (this.instance.getMajorVersion() != -1
                     && this.instance.getMajorVersion() != major) {    //changed < to !=
                 
                 throw new UnfullfilledRequirementsException(
                         "Major Version Mismatch (Required: " + major + " | Available: " + this.instance.getMajorVersion() + ")",
                         UnfullfilledRequirementsException.SEVERITY_PROMPT,
-                        UnfullfilledRequirementsException.REQUIREMENT_MAJOR_APP_VERSION,major,minor,this.instance.getMajorVersion(),this.instance.getMinorVersion());
+                        UnfullfilledRequirementsException.REQUIREMENT_MAJOR_APP_VERSION, major, minor, this.instance.getMajorVersion(), this.instance.getMinorVersion());
             }
             
-            //For the minor version, anything greater than the profile's version is valid
+            // For the minor version, anything greater than the profile's version is valid
             if (this.instance.getMinorVersion() != -1
                     && this.instance.getMinorVersion() < minor) {
                 throw new UnfullfilledRequirementsException(
                         "Minor Version Mismatch (Required: " + minor + " | Available: " + this.instance.getMinorVersion() + ")",
                         UnfullfilledRequirementsException.SEVERITY_PROMPT,
-                        UnfullfilledRequirementsException.REQUIREMENT_MINOR_APP_VERSION,major,minor,this.instance.getMajorVersion(),this.instance.getMinorVersion());
+                        UnfullfilledRequirementsException.REQUIREMENT_MINOR_APP_VERSION, major, minor, this.instance.getMajorVersion(), this.instance.getMinorVersion());
             }
         }
 
         String registrationNamespace = null;
-        
-        Profile profile = new Profile(version, authRef);
+
+        // If this is an old version of the profile file and is therefore missing uniqueId,
+        // get it from the update URL
+        boolean fromOld = false;
+        if (uniqueId == null) {
+            fromOld = true;
+            int startIndex = authRef.indexOf("download") + 9;
+            int endIndex = authRef.indexOf("profile") - 1;

```

Comment:
Paranoid, but perhaps this should use the indexOf(String str, int fromIndex) version and start from startIndex.